### PR TITLE
Support distribution to more than 3 dimension at the workgroup level

### PR DESCRIPTION
**Pull request by ThomasRaoux**
So far we were not distributing higher level dimensions for operations with more than 3 parallel dimensions. This delinearize the z dimension for case with more than 3 dimension and allow better distribution if the kernel config asks for it.
Also change the default config of llvmgpu and spirv backend to take advantage of that.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp [hunk 1]
```diff
@@ -41,9 +42,11 @@ llvm::SmallVector<unsigned> getPartitionableLoopsImpl(
   parallelLoops = pruneUnitTripParallelLoops(parallelLoops, staticLoopRanges);
   // TODO(ravishankarm): For now the outer parallel loops are dropped. This is
   // a pragmatic choice for now but might need to be revisited.
-  if (parallelLoops.size() > maxNumPartitionedLoops) {
-    parallelLoops = llvm::to_vector(llvm::ArrayRef<unsigned>(parallelLoops)
-                                        .take_back(maxNumPartitionedLoops));
+  if (maxNumPartitionedLoops.has_value() &&
+      parallelLoops.size() > maxNumPartitionedLoops.value()) {
+    parallelLoops =
+        llvm::to_vector(llvm::ArrayRef<unsigned>(parallelLoops)
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp:48]: Nit: You can use llvm::makeArrayRef

File: compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp [hunk 2]
```diff
@@ -99,11 +102,12 @@ struct OuterParallelAsPartitionableLoops
...
       partitionableLoops.push_back(iteratorType.index());
     }
-    if (partitionableLoops.size() > maxNumPartitionedLoops) {
-      partitionableLoops.erase(
-          partitionableLoops.begin(),
-          std::next(partitionableLoops.begin(),
-                    partitionableLoops.size() - maxNumPartitionedLoops));
+    if (maxNumPartitionedLoops.has_value() &&
+        partitionableLoops.size() > maxNumPartitionedLoops.value()) {
+      partitionableLoops.erase(partitionableLoops.begin(),
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp:107]: Style nit. Just the indentations gotten a bit unreadable. Maybe flip the condition, so
```
if (!maxNumPartitionableLoops.has_value() || partitionableLoops.size() <= maxNumPartitionedLoops.value()) {
  return partitionableLoops;
}
partitionableLoops.erase(...)

```

File: compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp [hunk 3]
```diff
@@ -125,19 +129,20 @@ struct FftOpPartitionableLoops
...
     // Indices matter for coeff computation.
     if (!fftOp.hasCoeff()) {
       partitionableLoops.pop_back();
     }
-    if (partitionableLoops.size() > maxNumPartitionedLoops) {
-      partitionableLoops.erase(
-          partitionableLoops.begin(),
-          std::next(partitionableLoops.begin(),
-                    partitionableLoops.size() - maxNumPartitionedLoops));
+    if (maxNumPartitionedLoops.has_value() &&
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp:140]: Same here . Style nit. Just the indentations gotten a bit unreadable. Maybe flip the condition, so
```
if (!maxNumPartitionableLoops.has_value() || partitionableLoops.size() <= maxNumPartitionedLoops.value()) {
  return partitionableLoops;
}
partitionableLoops.erase(...)

```

File: compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp [hunk 4]
```diff
@@ -160,11 +165,12 @@ struct AllParallelAsPartitionableLoops
       }
       partitionableLoops.push_back(iteratorType.index());
     }
-    if (partitionableLoops.size() > maxNumPartitionedLoops) {
-      partitionableLoops.erase(
-          partitionableLoops.begin(),
-          std::next(partitionableLoops.begin(),
-                    partitionableLoops.size() - maxNumPartitionedLoops));
+    if (maxNumPartitionedLoops.has_value() &&
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp:168]: (and here as well)
Style nit. Just the indentations gotten a bit unreadable. Maybe flip the condition, so
```
if (!maxNumPartitionableLoops.has_value() || partitionableLoops.size() <= maxNumPartitionedLoops.value()) {
  return partitionableLoops;
}
partitionableLoops.erase(...)

```

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 5]
```diff
@@ -633,13 +634,53 @@ static Value buildHALWorkgroupInfoOp(OpBuilder &b, unsigned dim) {
...
+        SmallVector<int64_t> nonZeroTileSizes;
+        for (int64_t size : tileSizes) {
+          if (size != 0) nonZeroTileSizes.push_back(size);
+        }
         auto numParallelDims = parallelLoopRanges.size();
 
         SmallVector<linalg::ProcInfo, 3> procInfo(numParallelDims);
+        Value splitDim;
         for (size_t dim = 0; dim < numParallelDims; ++dim) {
+          if (numParallelDims > 3 && dim >= 2) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:651]: Could we use `kNumMaxParallelDims` instead of hard coding it?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just a few minor comments. THanks for finishing this up. I had left this one piece undone, but good that you didnt hit any footguns. @dcaballe and @vmurali something to add to the CPU side as well.

**Comment by ThomasRaoux** [comment]
> Thanks Thomas implementing this after our discussion.
> 
> One small note, I think we should linearize the `X` dimension instead of the `Z`. Because in CUDA the limit of `X` is **2^31-1** while the limit of `Z` is **65535**. So X can cover larger iterations. As I understand it, the iree schedules `X` dimension to the innermost loops, maybe loop scheduling needs to be changed as well.

the problem is that X is the fastest moving dimension so we want to map it to the most inner tile sizes. Most likely the best solution would be to only use X and split it for all the dimension. This is probably too much of a change for this PR. Let’s discuss about changing that in the future.

> Another small note, when the trip count is known and very small (very small needs to be defined) maybe it's better to unroll the small loop. This way, there will be no need for generating instructions such as `rem` or `div`.

Right, this is a decision that should be taken by the configuration. The configuration decides on the tile size which effectively allows to control how much of the loop is unrolled vs distributed.

File: compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp [hunk 6]
```diff
@@ -41,9 +42,11 @@ llvm::SmallVector<unsigned> getPartitionableLoopsImpl(
   parallelLoops = pruneUnitTripParallelLoops(parallelLoops, staticLoopRanges);
   // TODO(ravishankarm): For now the outer parallel loops are dropped. This is
   // a pragmatic choice for now but might need to be revisited.
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.cpp:44]: Right, this should be done in each target config as we don't really hav target info there.

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
Awesome!!

**Pull request merged**